worksheet.rb

```diff
@@ -23,12 +56,98 @@
 # into this data structure, such as "DR0004"
 # and "3rd Feb 2016" and "RD0022"
 
-########################################################
+#######################################################
 # Step 4: Total Driver's Earnings and Number of Rides
-
 # Use an iteration blocks to print the following answers:
+
+def line
+  puts "#######################################################"
+end
+
+line
+
 # - the number of rides each driver has given
+
+def num_ride(ride_share)
+  ride_share.each do |key, value|
+    puts "#{key} has #{value.length} rides."
+    end
+end
+
+num_ride(ride_share)
+
+line
+
 # - the total amount of money each driver has made
+
+def driver_earn(ride_share)
+  ride_share.each do | driver, array |
+    total = 0
+     array.each do |ride|
+       total += ride[:cost]
+    end
+    puts "#{driver} made $#{total}"
+  end
+end
+
+driver_earn(ride_share)
+
+line
+
 # - the average rating for each driver
+
+def driver_avg(ride_share)
+  ride_share.each do | driver, array |
+    total = 0
+     array.each do |ride|
+       total += ride[:rating]
+     end
+    average = total.to_f / array.length
+    puts "#{driver} made #{average.round(1)}"
+  end
+end
+
+driver_avg(ride_share)
+
+line
+
 # - Which driver made the most money?
-# - Which driver has the highest average rating?
\ No newline at end of file
+
+def driver_max_earn(ride_share)
```

Comment:
Note that `num_ride`, `driver_earn`, and `driver_avg` all have this loop `ride_share.each do | driver, array |`. Consider how you could combine this functionality to reduce repeated code. It a delicate balance between writing many methods so that each method only does one thing and writing fewer methods to avoid repeated code.